### PR TITLE
Newapp: Remove hard coded image names for code detection

### PR DESCRIPTION
**Pull request by csrwng**
Removes hard coded image searcher for code detection
Adds resolver that uses "supports" annotation tags on images
Removes old code

### CODE REVIEW

File: examples/image-streams/image-streams-centos7.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
               "description": "Build and run Ruby 2.0 applications",
               "iconClass": "icon-ruby",
               "tags": "builder,ruby",
+              "supports": "ruby:2.0,ruby",
```
Comment (smarterclayton) [inline examples/image-streams/image-streams-centos7.json]: Why do you need supports "ruby:2.0"?  Shouldn't the lookup algorithm be to "ruby:2.0" automatically from the name of the tag and the name of the stream?  Or are you just being explicit?

**Comment by csrwng** [comment]
@smarterclayton - Need to finish the test for the new resolver, but wanted to get your take on it sooner than later. This needs to go in 3.0

File: pkg/generate/app/imagelookup.go [hunk 2]
```diff
@@ -349,3 +359,113 @@ func InputImageFromMatch(match *ComponentMatch) (*ImageRef, error) {
...
+					match := &ComponentMatch{
+						Value:       value,
+						Name:        stream.Name,
+						Description: fmt.Sprintf("Image stream %s in project %s, tracks %q", stream.Name, stream.Namespace, stream.Status.DockerImageRepository),
+						Builder:     IsBuilderImage(&imageData.DockerImageMetadata),
+						Score:       score,
+
+						ImageStream: &stream,
+						Image:       &imageData.DockerImageMetadata,
+						ImageTag:    imageapi.DefaultImageTag,
```
Comment (ncdc) [inline pkg/generate/app/imagelookup.go]: `tag`?

File: examples/image-streams/image-streams-centos7.json (hunk 1 above)
Comment (csrwng) [inline examples/image-streams/image-streams-centos7.json]: Hmm, I am being explicit, 'cause that's what I understood you said :-) https://github.com/openshift/origin/issues/1878

File: pkg/generate/app/imagelookup.go (hunk 2 above)
Comment (csrwng) [inline pkg/generate/app/imagelookup.go]: yes, thank you

File: examples/image-streams/image-streams-centos7.json (hunk 1 above)
Comment (smarterclayton) [inline examples/image-streams/image-streams-centos7.json]: Ok, be explicit *AND* do both checks.  I think the priority order should be "if you say ruby:2.0, we'll first check for IS ruby and tag 2.0, then search IS ruby for tag 2.0, then search other IS for them".  Is that what you're implementing (and by order I really mean weight)?

Comment (csrwng) [inline examples/image-streams/image-streams-centos7.json]: I don't get the difference between "we'll first check for IS ruby and tag 2.0, then search IS ruby for tag 2.0" isn't that the same? But yes, I'm using a weighted resolver to find the right image: https://github.com/openshift/origin/pull/3057/files#diff-67f42376bdba60c035c6268d160059baR268

Comment (smarterclayton) [inline examples/image-streams/image-streams-centos7.json]: Hrm, I'm rethinking this.  Say I have image stream:

ruby
  2.0.x - changes for 2.x
  2.0   - intended for exact

I want to control which thing picks 2.0.  So in that case, I want supports "ruby:2.0,ruby" on tag 2.0.x and I want clients to choose it.

So:

ruby:
  2.0.x: supports: "ruby:2.0"
  2.0: supports: ""

should pick ruby:2.0.x.  So having a supports tag should weight you higher than matching image stream name and tag name exactly.

----- Original Message -----

Comment (csrwng) [inline examples/image-streams/image-streams-centos7.json]: makes sense... I'll update the image stream resolver to weigh the supports annotation as well

File: pkg/generate/app/imagelookup.go [hunk 3]
```diff
@@ -349,3 +359,113 @@ func InputImageFromMatch(match *ComponentMatch) (*ImageRef, error) {
...
+
+// Resolve finds image stream images using their 'supports' annotation
+func (r *ImageStreamByAnnotationResolver) Resolve(value string) (*ComponentMatch, error) {
+	matches := []*ComponentMatch{}
+	for _, namespace := range r.Namespaces {
+		streams, err := r.getImageStreams(namespace)
+		if err != nil {
+			return nil, err
+		}
+		for _, stream := range streams {
```
Comment (smarterclayton) [inline pkg/generate/app/imagelookup.go]: This is way too deeply nested.

File: pkg/generate/app/cmd/newapp_test.go [hunk 4]
```diff
@@ -590,13 +554,54 @@ func TestRunAll(t *testing.T) {
 				"imageStream":      {"example", "ruby-hello-world"},
 				"buildConfig":      {"ruby-hello-world"},
 				"deploymentConfig": {"ruby-hello-world"},
+				"service":          {"ruby-hello-world"},
 			},
 			expectedErr:    nil,
 			expectInsecure: util.NewStringSet("example"),
 		},
+		{
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp_test.go]: Is there an annotation priority test case in here?

Comment (csrwng) [inline pkg/generate/app/cmd/newapp_test.go]: There isn't, adding one, will update the PR shortly

**Comment by csrwng** [comment]
updated with supports resolver tests

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**